### PR TITLE
Issue 88: 2. Add connections configuration

### PR DESCRIPTION
**Pull request by LourensVeen**
This factors out the Settings class into a separate module, and then adds a section for configuring container connections.

Please don't merge, the base branch is wrong because this is one of my famous stacked PRs.

### CODE REVIEW

File: mahiru/components/settings.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+from pathlib import Path
+from typing import AnyStr, IO, List, Optional, Union
+
+import ruamel.yaml as yaml
+import yatiml
+
+from mahiru.definitions.identifier import Identifier
+
+
+class AssetConnections:
```
Comment (svenvanderburg) [inline mahiru/components/settings.py]:
```suggestion
class AssetConnectionSettings:
```
?? To make it clear this is settings?

File: mahiru/components/settings.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    Attributes:
+        enabled: Whether to allow remote connections to assets.
+        external_ip: External IPv4 address we are available at.
+        ports: Port range to use for serving incoming connections,
+            as a length-2 list of the form [min, max].
+    """
+    def __init__(
+            self, enabled: bool = False, external_ip: Optional[str] = None,
+            ports: Optional[List[int]] = None) -> None:
+        """Create a ConnectionSettings object.
```
Comment (svenvanderburg) [inline mahiru/components/settings.py:23]: Still refer to ConnectionSettings

File: mahiru/components/settings.py (hunk 1 above)
Comment (svenvanderburg) [reply mahiru/components/settings.py]: Or you could refer to it as `settings.AssetConnections`

File: mahiru/components/settings.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+            if ports[0] > ports[1]:
+                raise RuntimeError('Minimum must be <= maximum')
+
+        self.enabled = enabled
+        self.external_ip = external_ip
+        self.ports = ports
+
+
+class Settings:
```
Comment (svenvanderburg) [inline mahiru/components/settings.py]:
```suggestion
class SiteSettings:
```

**Review by svenvanderburg — APPROVED** [review]
Love it! This is really clean now 👍 I left a few minor comments.

File: mahiru/components/settings.py (hunk 1 above)
Comment (LourensVeen) [reply mahiru/components/settings.py]: Yeah, I'm not very happy with the naming here either. Maybe `NetworkSettings`? That's shorter than `AssetConnectionSettings` but still covers the contents.

File: mahiru/components/settings.py (hunk 3 above)
Comment (LourensVeen) [reply mahiru/components/settings.py]: I like adding `Site`, and I think I might prefer `SiteConfiguration` actually. After all, it represents the entire config file, and that's what you usually call them.

**Comment by LourensVeen** [comment]
Okay, they're now SiteConfiguration and NetworkSettings. I'd like to hear what you think.

Also, are you sure you don't want to suggest that the settings should be a singleton/global object?

**Comment by svenvanderburg** [comment]
NetworkSettings​ sounds good!


[https://esciencecenter.nl/wp-content/themes/raadhuis/dist/assets/img/e-signature.png]
Sven van der Burg (he, him, his)
eScience Research Engineer/Instructor
My working days are Monday, Tuesday, Wednesday, and Thursday.
Netherlands eScience Center
www.esciencecenter.nl<https://www.esciencecenter.nl/>
________________________________

**Review by svenvanderburg — APPROVED** [review]

**Comment by svenvanderburg** [comment]
> Okay, they're now SiteConfiguration and NetworkSettings. I'd like to hear what you think.
> Also, are you sure you don't want to suggest that the settings should be a singleton/global object?
I thought I could get away with responding to Github emails only, but that clearly doesn't work!

I really like the new names!
And about singletons: In this case it seems like the SiteConfiguration and NetworkSettings are only used in the `Site` object (I only had a quick look though), which makes sense because you need those settings to setup the site correctly. So I would argue against a singleton. If it is (or becomes) more like a global configuration then indeed it would make sense to create a singleton.

**Comment by LourensVeen** [comment]
Ah, merged it but forgot to adjust the reference branch here, so GitHub didn't detect that. Closing.

**Pull request closed without merging**